opcodes/dxil/dxil_nvapi.cpp

```diff
@@ -258,6 +258,81 @@ static bool emit_nvapi_extn_op_get_special(Converter::Impl &impl)
 	return false;
 }
 
+static spv::Id bitcast_operand_to_float(Converter::Impl &impl, const LLVMBC::Value *op)
+{
+	auto *cast_inst = llvm::dyn_cast<llvm::CastInst>(op);
+
+	if (cast_inst && cast_inst->getOpcode() == llvm::Instruction::BitCast)
+	{
+		op = cast_inst->getOperand(0);
+
+		if (op->getType()->getTypeID() == llvm::Type::TypeID::FloatTyID)
+			return impl.get_id_for_value(op);
+	}
+
+	auto *bitcast_op = impl.allocate(spv::OpBitcast, impl.builder().makeFloatType(32));
+	bitcast_op->add_id(impl.get_id_for_value(op));
+	impl.add(bitcast_op);
+
+	return bitcast_op->id;
+}
+
+static bool emit_nvapi_extn_op_hit_object_make_miss(Converter::Impl &impl)
+{
+	spv::Id index;
+	if (!get_argument(impl, NVAPI_ARGUMENT_SRC0U + 0, &index))
+		return false;
+
+	auto *tmin = impl.nvapi.fake_doorbell_inputs[NVAPI_ARGUMENT_SRC0U + 1];
+	auto *tmax = impl.nvapi.fake_doorbell_inputs[NVAPI_ARGUMENT_SRC0U + 2];
+	auto *origin_x = impl.nvapi.fake_doorbell_inputs[NVAPI_ARGUMENT_SRC1U + 0];
+	auto *origin_y = impl.nvapi.fake_doorbell_inputs[NVAPI_ARGUMENT_SRC1U + 1];
+	auto *origin_z = impl.nvapi.fake_doorbell_inputs[NVAPI_ARGUMENT_SRC1U + 2];
+	auto *direction_x = impl.nvapi.fake_doorbell_inputs[NVAPI_ARGUMENT_SRC2U + 0];
+	auto *direction_y = impl.nvapi.fake_doorbell_inputs[NVAPI_ARGUMENT_SRC2U + 1];
+	auto *direction_z = impl.nvapi.fake_doorbell_inputs[NVAPI_ARGUMENT_SRC2U + 2];
+
+	if (!tmin || !tmax || !origin_x || !origin_y || !origin_z || !direction_x || !direction_y || !direction_z)
+		return false;
+
+	auto &builder = impl.builder();
+
+	builder.addExtension("SPV_NV_shader_invocation_reorder");
+	builder.addCapability(spv::CapabilityShaderInvocationReorderNV);
+
+	spv::Id t_min = bitcast_operand_to_float(impl, tmin);
+	spv::Id t_max = bitcast_operand_to_float(impl, tmax);
+
+	spv::Id ray_origin[3] = {
+		bitcast_operand_to_float(impl, origin_x),
+		bitcast_operand_to_float(impl, origin_y),
+		bitcast_operand_to_float(impl, origin_z),
+	};
+	spv::Id ray_dir[3] = {
+		bitcast_operand_to_float(impl, direction_x),
+		bitcast_operand_to_float(impl, direction_y),
+		bitcast_operand_to_float(impl, direction_z),
+	};
+
+	spv::Id float32 = builder.makeFloatType(32);
+	spv::Id ray_origin_vec = impl.build_vector(float32, ray_origin, 3);
+	spv::Id ray_dir_vec = impl.build_vector(float32, ray_dir, 3);
+
+	spv::Id variable = impl.create_variable(spv::StorageClassPrivate, builder.makeHitObjectNVType());
```

Comment:
why Private? Function seems like it'd make more sense to me.

Reply:
I admit I guessed which one would be more appropriate, based on

* how `OpTypeRayQueryKHR` is emitted, which shares similar behaviors and restrictions with `OpTypeHitObjectNV`, and
* how `glslang` emits SPIR-V from GLSL that uses `GL_NV_shader_invocation_reorder` extension and `hitObjectNV` variables.

But if you feel like `Function` is better then I'll just change it.

Reply:
> If Function will cause problems down the line

That I'm not sure about. I can certainly try to switch to Function and see how various samples and games react.

> But ray query objects are allocated with a global ID

From the comment in `emit_allocate_ray_query` I suppose this is how DXIL does it, right? It's pretty hard to reason about `NvHitObject` in HLSL or DXIL when the compiler treats like a common single member struct…

Anyway, I'll check how Function behaves.

Reply:
I suppose it works okay from some light testing.